netebpfext/net_ebpf_ext_xdp.c

```diff
@@ -685,4 +709,149 @@ net_ebpf_ext_layer_2_classify(
-}
\ No newline at end of file
+}
+
+/**
+ * @brief Build a xdp_md_t context for the eBPF program. This includes copying the packet data and
+ * metadata into a contiguous buffer and building an MDL chain for the same.
+ *
+ * @param[in] data_in The packet data.
+ * @param[in] data_size_in The size of the packet data.
+ * @param[in] context_in The context.
+ * @param[in] context_size_in The size of the context.
+ * @param[out] context The context to be passed to the eBPF program.
+ * @retval STATUS_SUCCESS The operation was successful.
+ * @retval STATUS_INVALID_PARAMETER One or more parameters are incorrect.
+ * @retval STATUS_NO_MEMORY Failed to allocate resources for this operation.
+ */
+static uint64_t
+_ebpf_xdp_context_create(
+    _In_reads_bytes_opt_(data_size_in) const uint8_t* data_in,
+    size_t data_size_in,
+    _In_reads_bytes_opt_(context_size_in) const uint8_t* context_in,
+    size_t context_size_in,
+    _Outptr_result_maybenull_ void** context)
+{
+    NTSTATUS status;
+    net_ebpf_xdp_md_t* new_context = NULL;
+    MDL* mdl_chain = NULL;
+    NET_BUFFER_LIST* new_nbl = NULL;
+
+    if (!context) {
+        status = STATUS_INVALID_PARAMETER;
+        goto Done;
+    }
+
+    *context = NULL;
+
+    // Data is mandatory.
+    // Context is optional.
+    if (data_in == NULL || data_size_in == 0) {
+        status = STATUS_INVALID_PARAMETER;
+        goto Done;
+    }
+
+    new_context = (net_ebpf_xdp_md_t*)ExAllocatePoolUninitialized(
+        NonPagedPoolNx, sizeof(net_ebpf_xdp_md_t), NET_EBPF_EXTENSION_POOL_TAG);
+    if (new_context == NULL) {
+        status = STATUS_NO_MEMORY;
+        goto Done;
+    }
+
+    memset(new_context, 0, sizeof(net_ebpf_xdp_md_t));
+
+    // Create a MDL with the packet buffer.
+    mdl_chain = IoAllocateMdl((void*)data_in, (ULONG)data_size_in, FALSE, FALSE, NULL);
+    if (mdl_chain == NULL) {
+        status = STATUS_INSUFFICIENT_RESOURCES;
+        goto Done;
+    }
+    MmBuildMdlForNonPagedPool(mdl_chain);
+
+    // Now allocate the cloned NBL using this MDL chain.
+    status = FwpsAllocateNetBufferAndNetBufferList(
+        _net_ebpf_ext_nbl_pool_handle, 0, 0, mdl_chain, 0, data_size_in, &new_nbl);
+    if (!NT_SUCCESS(status))
+        goto Done;
+    mdl_chain = NULL;
+
+    new_context->original_nbl = new_nbl;
+    new_nbl = NULL;
+
+    new_context->base.data = (void*)data_in;
+    new_context->base.data_end = (void*)(data_in + data_size_in);
+
+    if (context_in != NULL && context_size_in >= sizeof(xdp_md_t)) {
+        xdp_md_t* xdp_context = (xdp_md_t*)context_in;
+        new_context->base.data_meta = xdp_context->data_meta;
+        new_context->base.ingress_ifindex = xdp_context->ingress_ifindex;
+    }
+
+    *context = new_context;
+    new_context = NULL;
+
+Done:
+    if (new_context) {
+        ExFreePool(new_context);
+        new_context = NULL;
+    }
+
+    if (mdl_chain) {
+        IoFreeMdl(mdl_chain);
+        mdl_chain = NULL;
```

Comment:
Explicitly setting ```mdl_chain``` (and ```new_nbl``` further below) to NULL _after_ we've already freed them is not required as we're not dependent on their values thereafter.

Reply:
My view: it's still fine practice.   If compiler optimization is on, it won't make any difference.  If compiler optimization is off, then debugging won't try to show you bytes that have been freed, if you null it out.  So I think this is ok.